### PR TITLE
Embed binaries

### PR DESCRIPTION
**Pull request by cbbruss**
Added functionality to embed a binary once you have a pre-trained model. It pads according to the maximum size of the batch so it to sorts by size first. 

I also added unit tests for everything, docstrings and some type hints.

### CODE REVIEW

File: byte_embeddings.py [hunk 1]
```diff
@@ -2,95 +2,77 @@
...
+        model = Word2Vec(size=256, window=2,
+                         workers=8, sg=1, hs=0,
+                         negative=5, min_count=0)
         model.build_vocab(b_words)
-    model.train(b_words, total_examples=len(b_words), epochs=15, compute_loss=True)
+    model.train(b_words, total_examples=len(b_words),
+                epochs=15, compute_loss=True)
     return model
 
-def generate_batch(zip_file, file_names, cur_idx, batch_size, password=None):
```
Comment (zhanif3) [inline byte_embeddings.py:49]: NOTE: No longer generating batches here, and doing them in a separate library.

File: byte_embeddings.py [hunk 2]
```diff
@@ -99,7 +81,8 @@ def generate_batch(zip_file, file_names, cur_idx, batch_size, password=None):
         print('Elapsed Time:', round(t_delt, 3))
 
         remaining_batches = (num_files - idx + batch_size) / batch_size
-        print('Estimated Time Remaining (hrs):', round(t_delt * remaining_batches / 3600))
+        rt = round(t_delt * remaining_batches / 3600)
```
Comment (zhanif3) [inline byte_embeddings.py:84]: I assume this was done for readability?

File: byte_embeddings.py [hunk 3]
```diff
@@ -2,95 +2,77 @@
...
+    parser.add_argument('--existing_model',
+                        default=None,
                         help='number of binaries to add to model at a time')
-    parser.add_argument('--existing_model', default=None, help='number of binaries to add to model at a time')
     args = parser.parse_args()
 
-    zip_file = ZipFile(args.zip_file_name, 'r')
-
-    fn = get_file_names(zip_file)
+    zdh = ZipDataHandler(args.zip_file_name, password=args.pass_string)
```
Comment (zhanif3) [inline byte_embeddings.py:55]: NOTE: lul.

File: byte_embeddings.py [hunk 4]
```diff
@@ -2,95 +2,77 @@
...
 
     batch_size = int(args.batch_size)
-    num_files = len(fn)
+    num_files = zdh.num_files
     print("Number of files: ", num_files)
     for idx in range(0, num_files, batch_size):
         st = time.time()
 
-        batch = generate_batch(zip_file, fn, idx, batch_size, args.pass_string)
+        batch = zdh.generate_batch(idx, batch_size)
```
Comment (zhanif3) [inline byte_embeddings.py:72]: What is the relationship between byte_embeddings.py and embed_binary.py? Comment suggests that one is done over bytecode, but this is not strictly accurate (we're training over raw bytes). One also sorts files by size before generating batches, which suggests that embed_binary represents the model as most recently discussed, and byte_embeddings is an older file that is retained in the source tree for historical reasons. Is this correct, or am I missing something key to why one gets size sorted and the other does not?

File: embed_binary.py [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+    zdh = ZipDataHandler(args.zip_file_name, password=args.pass_string)
+
+    zdh.sort_files_by_size()
+
+    w2v_model = Word2Vec.load(args.existing_model)
+
+    batch_size = int(args.batch_size)
+    num_files = zdh.num_files
+    print("Number of files: ", num_files)
+    embedding_dim = len(w2v_model.wv['00e6'])
```
Comment (zhanif3) [inline embed_binary.py]: What's with the magic string? Why was that particular byte sequence selected?

File: byte_embeddings.py (hunk 1 above)
Comment (cbbruss) [reply byte_embeddings.py:49]: Yeah, I pulled it out into the zip data handler class in util_functions.py

File: byte_embeddings.py (hunk 2 above)
Comment (cbbruss) [reply byte_embeddings.py:84]: yeah, flake8 is unforgiving.

File: byte_embeddings.py (hunk 4 above)
Comment (cbbruss) [reply byte_embeddings.py:72]: Sorry for the confusion on comment. So basically byte_embeddings takes a bunch of raw files and uses them as training data for a model.  Once that model has been trained we need to go back into the binaries and replace all the hex with their corresponding embeddings from the trained model. This is what embed_binary does.

File: embed_binary.py (hunk 5 above)
Comment (cbbruss) [reply embed_binary.py]: that was me being lazy and not figuring out how to request the emebdding size from the model class itself, I am sure there is a method, I'll go and update it.

**Comment by cbbruss** [comment]
@zhanif3 is this ok? I am going to merge it an open a new one for reducing a binary to single vector.

**Comment by zhanif3** [comment]
Sorry, LGTM.

**Review by zhanif3 — APPROVED** [review]

**Pull request merged**